### DISCUSSION TITLE
3000 documents too much for Docusaurus?

### DISCUSSION THREAD

**Discussion by osseonews** [facebook/docusaurus#10159] [Q&A]
Hi, I'm wondering if 3,000 documents would be too much to handle for Docusaurus on build? How long do you think such a build would take and is there some sort of upper limit of docs where docusasurus SSG would just not be worth it because the build time would be too long?

**Comment by OzakIOne** [comment]
There is a similar issue https://github.com/facebook/docusaurus/issues/7256 with people having some issues building thousand of documents with Docusaurus.

There is also something here https://github.com/facebook/docusaurus/issues/4765 

> In our case (a site with around 2,000 .md(x) files) most of the time spent seems to be before and after the "Compiling Client/Compiling Server" progress bars appear and complete.
> As it stands, building the site takes around 20 minutes with esbuild, and was closer to 40 minutes before.

These were with older versions of Docusaurus so it may be different now, and it also depends if you want to have different language for your site, and it may take a bit longer if you use specific features

**Comment by slorber** [comment]
We don't really have any estimate, it highly depends on what's on your pages, the computer/memory/CPU you have, what you consider "too long" etc... The best is to do a POC is 3000x content docs that are representative of the site you plan to build, and see if it's acceptable. 

We'll continue to try improving Docusaurus performance (dev, build time and runtime) and post results in releases notes and in https://github.com/facebook/docusaurus/issues/4765

**Comment by osseonews** [comment]
Thanks for the answers. We use Cloudflare Pages to host Docusaurus. Do you think we will benefit if we just build the site with their Direct  Upload feature? https://developers.cloudflare.com/pages/get-started/direct-upload/. That way we should be able to take advantage of the Docusaurus incremental builds, otherwise Cloudflare reinstalls everything and doesn't persist node_modules/.cache . Also, I saw this new feature on Cloudflare, https://developers.cloudflare.com/pages/configuration/build-caching/, but didn't see Docusaurus as a framework that was supported for Build Caching. Maybe you can suggest it? I think the Build Cache would make much easier to persist the cache, as Direct Upload is a bit of a pain.

**Reply by slorber** [reply]
If Cloudflare can't persist Docusaurus caches across build, it's probably simpler to use an alternative solution like Vercel/Netlify instead.

I just suggested them on Discord to add Docusaurus to that list.
However are you sure incremental builds don't work? Because it looks like as part of dependency caching, `node_modules/.cache` is in `node_modules` so it could already be cached?

**Reply by dario-piotrowicz** [reply]
@slorber no, those directories are not cached 🙂 

I'll look into adding the build cache support as discussed on discord 👍 

The directories to be cached are:
- `/node_modules/.cache`
-  `/.docusaurus`
-  `/build` 

right?

**Reply by osseonews** [reply]
Thanks for reaching out to Cloudflare. We make heavy use of Cloudflare functions(workers), Cloudflare KV and other Cloudflare services to create a more "interactive" Docusaurus site. So it's kind of a necessity to host the whole application there. Honestly, we've moved away from other hosting solutions,  as Cloudflare has just so many great integrated services now that are owned/managed by Cloudflare - it's just a more powerful and easier to maintain solution.

**Reply by slorber** [reply]
Thanks @dario-piotrowicz 

Yes those 3 folders would be great. We are only using `/node_modules/.cache` atm but might use the other 2 in the future so it's nice to add them already.

**Reply by dario-piotrowicz** [reply]
quick update: I've made the changes internally, they need to be reviewed and released, I've also updated our docs: https://github.com/cloudflare/cloudflare-docs/pull/14761

**Reply by dario-piotrowicz** [reply]
@slorber the change has been released the the Cloudflare docs updated accordingly 👍 

You and/or @osseonews please have a look and let me know if it works well for you 🙂🙏

**Reply by slorber** [reply]
@dario-piotrowicz I'm not sure it's working, I tried to "retry" a deployment, and push a new empty commit, and in both cases the 2nd run doesn't really build faster. Locally, the project builds 4x faster on 2nd run.

I don't have a better way to troubleshoot this unfortunately, and I'm not sure Cloudflare Pages logs let me troubleshoot this either 😅 

- Cloudflare project: https://dash.cloudflare.com/4513c352aecf17e4a7c00de4bdb8dc1d/pages/view/docusaurus-starter
- Url: https://docusaurus-starter-3l0.pages.dev/
- Git project https://github.com/slorber/docusaurus-starter

```
yarn clear
yarn build
yarn build
```

The 2nd build should be significantly faster.

---

Note: this might be due to a Yarn issue? It seems you force use a v2+ version of Yarn while the project is using Yarn v1 lockfile. I had to use `YARN_ENABLE_IMMUTABLE_INSTALLS=false` to avoid this error:

![CleanShot 2024-06-17 at 18 48 18](https://github.com/facebook/docusaurus/assets/749374/c2f48cac-ebb6-4a0a-9911-5dca07ee3a56)

**Reply by dario-piotrowicz** [reply]
@slorber thanks for letting me know! (and for the various resources/links you've shared that will surely come in handy! ❤️)

I did configure docusaurus in the standard way, so it should definitely work, but given the fact that the docusaurus build command itself doesn't provide logs for the caching I didn't fully test it e2e, so I think it should work but something might be going wrong, I'll have a look 👍 (if I ever loose track of this amidst other things please feel free to always ping me and remind me! 😅)

In the meantime, a couple of things:
 - are you using the [v2 build image](https://blog.cloudflare.com/moderizing-cloudflare-pages-builds-toolbox#new-defaults-new-possibilities)? (v1 does not support build caching) (I am pretty sure you do since you said you've seen the CI use yarn v2+ but I'm asking just to be extra extra sure)
 - could you try to build your app with other pms and see if that helps? (`npm` and `pnpm` in particular) 
 - you can ask Pages to use any version of yarn by setting the `YARN_VERSION` env variable ([docs](https://developers.cloudflare.com/pages/configuration/language-support-and-tools/#supported-languages-and-tools)), maybe you could try using v1 then?

**Reply by slorber** [reply]
@dario-piotrowicz looks like I'm using build image v2 yeah

![CleanShot 2024-06-18 at 17 05 13](https://github.com/facebook/docusaurus/assets/749374/b29bb898-af04-4481-b677-bb8028716943)

I turned Yarn 1 on with the env variable, and it still isn't faster on retry.
- Initial run: https://dash.cloudflare.com/4513c352aecf17e4a7c00de4bdb8dc1d/pages/view/docusaurus-starter/7723b939-0815-4037-aa48-f8fde8f58e99
- Second run: https://dash.cloudflare.com/4513c352aecf17e4a7c00de4bdb8dc1d/pages/view/docusaurus-starter/d6ed0a0f-e7ed-4125-abc8-6128a7f66501

I'm not sure how to try with other pms, I didn't ask for Yarn in the first place, it automatically decided to use it (probably based on the presence of `yarn.lock`)

I believe I chose the Docusaurus framework, but can't see anywhere to double-check I selected it properly. Maybe my initial choice wasn't applied properly?

On the Docusaurus side, there's no need to "print" that caching is on because caching is always on. The thing is, is the cache empty or not? It is webpack 5 persistent caching. Maybe we could get Webpack to log something about its caching system, but I don't know how. And it's unlikely to be "on/off", because it caches things with granularity and is able to ignore/invalidate a subset of its cache depending on files that have been updated.

**Reply by slorber** [reply]
Oh, 🤦 didn't know I had to turn on build caching 😅 now it works thanks

![CleanShot 2024-06-18 at 17 17 18](https://github.com/facebook/docusaurus/assets/749374/5e78de69-35a3-436a-961a-720753aae56a)

**Reply by dario-piotrowicz** [reply]
> On the Docusaurus side, there's no need to "print" that caching is on because caching is always on.

No, that's not what I meant, what I was referring to is that other frameworks that i integrated with the Pages build caching do have options so that the frameworks logs cache hits, misses and whatnot that's what I was thinking would be very valuable for validating what's happening here 🙂

**Reply by dario-piotrowicz** [reply]
> Oh, 🤦 didn't know I had to turn on build caching 😅 now it works thanks

lol... ok sorry I totally took that for granted! 😅 (it also bit me in the past too! 🤣)
(sorry for not mentioning this! 🙇)

Nice! so hopefully now it's all working nicely? 😃

**Reply by osseonews** [reply]
thanks very much for implementing this. By the way, I just tried it and it seems to work. Seems to have cut the build time down by nearly 40% when using the cache.  I was wondering, does the cache apply to the preview branches also or just the production branch?

**Reply by slorber** [reply]
Looks like it's working yes

Regarding logs, it's not Docusaurus that knows if there's a cache hit/miss, it's webpack, and I don't know how to inspect that 😅 

@osseonews from Docusaurus perspective it doesn't matter if it's a preview branch. As long as Cloudflare is able t cache per-branch (I don't know 🤷‍♂️ ) it should work.
